utils/include/detray/simulation/event_generator/random_track_generator.hpp

```diff
@@ -16,48 +16,74 @@
 
 // System include(s)
 #include <algorithm>
-#include <cmath>
+#include <array>
 #include <random>
 
 namespace detray {
 
-/// Wrapper for random number generatrion for the @c random_track_generator
+/// Wrapper for CPU random number generatrion for the @c random_track_generator
 template <typename scalar_t = scalar,
           typename distribution_t = std::uniform_real_distribution<scalar_t>,
-          typename generator_t = std::random_device,
           typename engine_t = std::mt19937_64>
 struct random_numbers {
 
-    random_numbers(random_numbers&& other) : engine(std::move(other.engine)) {}
-
-    generator_t gen;
-    engine_t engine;
-
-    template <
-        typename T = generator_t,
-        std::enable_if_t<std::is_same_v<T, std::random_device>, bool> = true>
-    random_numbers() : gen{}, engine{gen()} {}
-
-    template <typename T = generator_t,
-              std::enable_if_t<std::is_same_v<T, std::seed_seq>, bool> = true>
-    random_numbers() : gen{42u}, engine{gen} {}
-
-    template <typename T = distribution_t,
-              std::enable_if_t<
-                  std::is_same_v<T, std::uniform_real_distribution<scalar_t>>,
-                  bool> = true>
-    DETRAY_HOST auto operator()(const scalar_t min, const scalar_t max) {
-        return distribution_t(min, max)(engine);
+    using distribution_type = distribution_t;
+    using engine_type = engine_t;
+    using seed_type = typename engine_t::result_type;
+
+    std::seed_seq m_seeds;
+    engine_t m_engine;
+
+    /// Default seed
+    DETRAY_HOST
+    random_numbers()
+        : m_seeds{random_numbers::default_seed()}, m_engine{m_seeds} {}
+
+    /// Different seed @param s for every instance
+    DETRAY_HOST
+    random_numbers(seed_type s) : m_seeds{s}, m_engine{m_seeds} {}
+
+    /// More entropy in seeds from collection @param s
+    DETRAY_HOST
+    random_numbers(const std::vector<seed_type>& s)
+        : m_seeds{s.begin(), s.end()}, m_engine{m_seeds} {}
+
+    /// Copy constructor
+    DETRAY_HOST
+    random_numbers(random_numbers&& other)
+        : m_engine(std::move(other.m_engine)) {}
+
+    /// Generate random numbers in a given range
+    DETRAY_HOST auto operator()(const std::array<scalar_t, 2> range = {
+                                    -std::numeric_limits<scalar_t>::max(),
+                                    std::numeric_limits<scalar_t>::max()}) {
+        const scalar_t min{range[0]}, max{range[1]};
+        assert(min <= max);
+
+        // Uniform
+        if constexpr (std::is_same_v<
+                          distribution_t,
+                          std::uniform_real_distribution<scalar_t>>) {
+
+            return distribution_t(min, max)(m_engine);
+
+            // Normal
+        } else if constexpr (std::is_same_v<
```

Comment:
Why don't you just take range as mu and sigma directly for Gauss?

Reply:
I kept this as it was before. This is done to only cut away the tails from the Gaussian profile over the range, i.e. to not end up with a strongly truncated distribution. We should probably not call it "normal", but "Gaussian" or smth

Reply:
Yeah so why don't you change it more precise way? I think changing the name of `range` to `param` will give you better sense

Reply:
I don't think I understand what you mean? There already is a method called ```normal``` that does that. The random numbers in this method should adapt the values drawn from the distribution to ```min```, ```max``` instead

Reply:
what is the method of `normal`?

Reply:
```normal``` returns a random number drawn from a Gaussian according to a mean and stddev, which are passed as arguments, independent on whether or not the ```random_numbers``` type is set to use a uniform or Gaussian distribution